File: Data Structure/Array or Matrix/Easy/LET_no509_Fibonacci Number.py
class Solution:
    def fib(self, n: int) -> int:
        a0, a1 = 0, 1
        mat = self.fast_power(n)
        return mat[1][0] * a1 + mat[1][1] * a0
    
    def fast_power(self, n):
        base = [[1, 1], [1, 0]]
        res = [[1, 0], [0, 1]]
        while n:
            if n & 1:
                res = self.mat_p(res, base)
            base = self.mat_p(base, base)
            n //= 2
        return res
                
    def mat_p(self, a, b):
        temp = [[0, 0], [0, 0]]
        for i in range(2):
            for j in range(2):
                for k in range(2):
                    temp[i][j] += a[i][k] * b[k][j]
        return temp

File: Data Structure/Array or Matrix/Easy/LET_no509_Fibonacci Number.cpp
class Solution {
public:
    int fib(int n) {
        int a0 = 0, a1 = 1;
        vector<vector<int>> base = {{1, 1}, {1, 0}};
        vector<vector<int>> res = {{1, 0}, {0, 1}};
        fast_power(res, base, n);
        return res[1][0] * a1 + res[1][1] * a0;
    }
    
    void fast_power(vector<vector<int>>& res, vector<vector<int>>& base, int n) {
        vector<vector<int>> temp(2, vector<int>(0, 2));
        while (n) {
            if (n & 1) {
                temp = res;
                mat_p(res, temp, base);
            }
            temp = base;
            mat_p(base, temp, temp);
            n >>= 1;
        }
    }
    
    void mat_p(vector<vector<int>>& a, vector<vector<int>>& b, vector<vector<int>>& c) {
        for (int i = 0; i < 2; i++) {
            for (int j = 0; j < 2; j++) {
                a[i][j] = 0;
                for (int k = 0; k < 2; k++)
                    a[i][j] += b[i][k] * c[k][j];
            }
        }
    }
};